pkg/inclusion/nmt_caching.go

```diff
@@ -0,0 +1,125 @@
+	return &subTreeRootCacher{cache: make(map[string][2]string)}
+}
+
+// Visit fullfills the nmt.VisitorNode function definition. It stores each inner
+// node in a simple map, which can later be used to walk the tree. This function
+// is called by the nmt when calculating the root.
+func (strc *subTreeRootCacher) Visit(hash []byte, children ...[]byte) {
+	switch len(children) {
+	case 2:
+		strc.cache[string(hash)] = [2]string{string(children[0]), string(children[1])}
+	case 1:
+		return
+	default:
+		panic("unexpected visit")
+	}
+}
+
+// walk recursively traverses the subTreeRootCacher's internal tree by using the
+// provided sub tree root and path. The provided path should be a []bool, false
+// indicating that the first child node (left most node) should be used to find
+// the next path, and the true indicating that the second (right) should be used.
+// walk throws an error if the sub tree cannot be found.
+func (strc subTreeRootCacher) walk(root []byte, path []bool) ([]byte, error) {
+	// return if we've reached the end of the path
+	if len(path) == 0 {
+		return root, nil
+	}
+	// try to lookup the provided sub root
+	children, has := strc.cache[string(root)]
+	if !has {
+		// note: we might want to consider panicing here
+		return nil, fmt.Errorf("did not find sub tree root: %v", root)
+	}
+
+	// continue to traverse the tree by recursively calling this function on the next root
+	switch path[0] {
+	// walk left
+	case false:
+		return strc.walk([]byte(children[0]), path[1:])
+	// walk right
+	case true:
+		return strc.walk([]byte(children[1]), path[1:])
+	default:
+		// this is unreachable code, but the compiler doesn't recognize this somehow
+		panic("bool other than true or false, computers were a mistake, everything is a lie, math is fake.")
+	}
+}
+
+// EDSSubTreeRootCacher caches the inner nodes for each row so that we can
+// traverse it later to check for message inclusion. NOTE: Currently this has to
+// use a leaky abstraction (see docs on counter field below), and is not
+// threadsafe, but with a future refactor, we could simply read from rsmt2d and
+// not use the tree constructor which would fix both of these issues.
+type EDSSubTreeRootCacher struct {
+	caches     []*subTreeRootCacher
+	squareSize uint64
+	// counter is used to ignore columns NOTE: this is a leaky abstraction that
+	// we make because rsmt2d is used to generate the roots for us, so we have
+	// to assume that it will generate a row root every other tree contructed.
+	// This is also one of the reasons this implementation is not thread safe.
+	// Please see note above on a better refactor.
+	counter int
+}
+
+func NewCachedSubtreeCacher(squareSize uint64) *EDSSubTreeRootCacher {
+	return &EDSSubTreeRootCacher{
+		caches:     []*subTreeRootCacher{},
+		squareSize: squareSize,
+	}
+}
+
+// Constructor fullfills the rsmt2d.TreeCreatorFn by keeping a pointer to the
+// cache and embedding it as a nmt.NodeVisitor into a new wrapped nmt. I only
```

Comment:
[blocking]
the rest of this comment got truncated

Reply:
sorry, this was fixed two weeks ago, but I didn't realize github wasn't posting the comments